File: src/ollama_cli/blobs.py
#!/usr/bin/env python3
import json
from pathlib import Path
from typing import Any

from . import utils


# ============================================================
#  Model name cleaning
# ============================================================
def normalize_model_name(model_path: str) -> str:
    """Convert manifest path to a short model name (strip library/, cut at first /)."""
    if model_path.startswith("library/"):
        model_path = model_path[len("library/") :]
    return model_path.split("/", 1)[0]


# ============================================================
#  Blob → models mapping (from manifests)
# ============================================================
def collect_blob_mappings(manifest_root: Path) -> dict[str, set[str]]:
    """Extract blob -> models mapping from manifest files.
    
    Returns:
        blob_to_models: blob_hash -> set(models)
    """
    blob_to_models: dict[str, set[str]] = {}

    if not manifest_root.exists():
        return blob_to_models

    for mf in manifest_root.rglob("*"):
        if not mf.is_file():
            continue

        try:
            data = json.loads(mf.read_text(encoding="utf-8"))
        except Exception:
            continue

        raw_model_path = str(mf.relative_to(manifest_root))
        model_name = normalize_model_name(raw_model_path)

        def add_digest(digest: str | None, model_name=model_name):
            if isinstance(digest, str) and digest.startswith("sha256:"):
                blob_hash = digest.split("sha256:", 1)[1]
                blob_to_models.setdefault(blob_hash, set()).add(model_name)

        cfg = data.get("config")
        if isinstance(cfg, dict):
            add_digest(cfg.get("digest"))

        for layer in data.get("layers", []):
            if isinstance(layer, dict):
                add_digest(layer.get("digest"))

    return blob_to_models


# ============================================================
#  Ensure *all* files in blobs directory are processed
# ============================================================
def extract_blob_hash_relaxed(name: str) -> str | None:
    """Very relaxed hash extraction for filenames that do NOT start with 'sha256-'.

    Used as a fallback only; any 'sha256-...' file is always included directly.
    """
    n = name.lower()
    n = n.split(".", 1)[0]  # strip extensions

    if n.startswith("sha256"):
        n = n[len("sha256") :]

    if len(n) < 40 or len(n) > 128:
        return None
    if not all(c in "0123456789abcdef" for c in n):
        return None

    return n


def list_all_blobs(blobs_root: Path) -> list[str]:
    """Return ALL blob hashes found for files in blobs_root.

    Rules:
    - If filename starts with 'sha256-': we *always* take everything after 'sha256-'
      as the hash, no validation.
    - Otherwise, we try a relaxed extractor which attempts to parse a hex-ish hash.
    """
    blobs: set[str] = set()

    if not blobs_root.exists():
        return []

    for f in blobs_root.iterdir():
        if not f.is_file():
            continue
        name = f.name

        if name.startswith("sha256-") and len(name) > len("sha256-"):
            blobs.add(name[len("sha256-") :])
            continue

        h = extract_blob_hash_relaxed(name)
        if h:
            blobs.add(h)

    return sorted(blobs)


# ============================================================
#  Blob analysis command
# ============================================================
def cmd_blobs(args: Any) -> None:
    """Analyze blobs and manifests, find orphans."""
    root = Path(args.models_root).expanduser()
    manifest_root = root / "manifests" / "registry.ollama.ai"
    blobs_root = root / "blobs"

    if args.debug:
        print(f"DEBUG: models_root   = {root}")
        print(f"DEBUG: manifest_root = {manifest_root}")
        print(f"DEBUG: blobs_root    = {blobs_root}")

    blob_to_models = collect_blob_mappings(manifest_root)
    blobs = list_all_blobs(blobs_root)

    if args.debug:
        print(f"DEBUG: total blobs found in blobs/ = {len(blobs)}")
        if args.debug_blob:
            hb = args.debug_blob.lower()
            in_list = any(b.startswith(hb) for b in blobs)
            print(f"DEBUG: debug-blob '{hb}' present in blobs list: {in_list}")

    rows: list[dict] = []
    orphan_files: list[Path] = []

    total_blobs = len(blobs)
    if args.progress and total_blobs > 0:
        print(f"Processing {total_blobs} blobs...")

    for idx, blob_hash in enumerate(blobs, start=1):
        if args.progress and total_blobs > 0:
            utils.progress_bar(idx, total_blobs, prefix="Blobs: ")

        blob_file = blobs_root / f"sha256-{blob_hash}"
        models = sorted(blob_to_models.get(blob_hash, []))
        bsize = utils.size_bytes(blob_file)
        size_str = utils.format_size(bsize, args.format)

        is_orphan = len(models) == 0
        if is_orphan:
            orphan_files.append(blob_file)

        rows.append(
            {
                "blob": f"sha256-{blob_hash}",
                "models": "|".join(models),
                "size_bytes": bsize,
                "size": size_str,
                "is_orphan": "yes" if is_orphan else "no",
            }
        )

    if args.progress and total_blobs > 0:
        print()  # newline after progress bar

    # filter only-orphans
    if args.only_orphans:
        rows = [r for r in rows if r["is_orphan"] == "yes"]

    # sorting
    reverse_sort = args.sort_desc
    if args.sort_by_blob:
        rows.sort(key=lambda r: r["blob"], reverse=reverse_sort)
    elif args.sort_by_model:
        rows.sort(
            key=lambda r: (r["models"] == "", r["models"], r["blob"]),
            reverse=reverse_sort,
        )
    elif args.sort_by_size:
        rows.sort(key=lambda r: r["size_bytes"], reverse=reverse_sort)

    # columns
    all_cols = {"blob", "models", "size_bytes", "size", "is_orphan"}
    columns = [c.strip() for c in args.columns.split(",") if c.strip()]
    if not columns:
        import sys
        print("Error: no columns specified", file=sys.stderr)
        sys.exit(2)
    unknown = [c for c in columns if c not in all_cols]
    if unknown:
        import sys
        print(
            f"Error: unknown columns: {', '.join(unknown)}; "
            f"allowed: {', '.join(sorted(all_cols))}",
            file=sys.stderr,
        )
        sys.exit(2)

    # output
    if args.as_csv:
        utils.write_csv(rows, columns, args.output)
    else:
        utils.print_table(
            rows, columns, enable_color=(not args.no_color and utils.supports_color())
        )

    # delete orphans
    if args.delete_orphans:
        if not orphan_files:
            print("\nNo orphan blobs found.")
            return

        print("\nOrphan blobs to delete:")
        for f in orphan_files:
            print("  blobs/" + f.name)

        if not args.force:
            confirm = input("Delete? (yes/no) ").strip().lower()
            if confirm != "yes":
                print("Aborted.")
                return

        for f in orphan_files:
            try:
                f.unlink()
                print("Deleted: blobs/" + f.name)
            except Exception as e:
                print("Failed:", f, e)

        print("Done.")


# ============================================================
#  models aggregation command
# ============================================================
def cmd_models(args: Any) -> None:
    """Show models with total blob size and blob count, based on manifests + blobs."""
    root = Path(args.models_root).expanduser()
    manifest_root = root / "manifests" / "registry.ollama.ai"
    blobs_root = root / "blobs"

    blob_to_models = collect_blob_mappings(manifest_root)
    blobs = list_all_blobs(blobs_root)

    # Precompute blob sizes
    blob_sizes: dict[str, int] = {}
    for blob_hash in blobs:
        blob_file = blobs_root / f"sha256-{blob_hash}"
        blob_sizes[blob_hash] = utils.size_bytes(blob_file)

    # Aggregate per model
    model_info: dict[str, dict[str, object]] = {}
    for blob_hash, models in blob_to_models.items():
        size_b = blob_sizes.get(blob_hash, 0)
        for m in models:
            info = model_info.setdefault(
                m, {"model": m, "blob_count": 0, "total_bytes": 0}
            )
            info["blob_count"] = int(info["blob_count"]) + 1
            info["total_bytes"] = int(info["total_bytes"]) + size_b

    # Convert to rows
    rows: list[dict] = []
    for m, info in model_info.items():
        total_b = int(info["total_bytes"])
        rows.append(
            {
                "model": m,
                "blob_count": int(info["blob_count"]),
                "total_bytes": total_b,
                "total_size": utils.format_size(total_b, args.format),
            }
        )

    # sorting
    reverse = args.sort_desc
    if args.sort_by_size:
        rows.sort(key=lambda r: r["total_bytes"], reverse=reverse)
    else:
        rows.sort(key=lambda r: r["model"], reverse=reverse)

    columns = ["model", "blob_count", "total_size", "total_bytes"]
    if args.as_csv:
        utils.write_csv(rows, columns, args.output)
    else:
        utils.print_table(
            rows, columns, enable_color=(not args.no_color and utils.supports_color())
        )

File: go/cmd/blobs.go
package cmd

import (
	"encoding/csv"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/spf13/cobra"
	"github.com/r14r/ollama-cli/internal/blobscan"
	"github.com/r14r/ollama-cli/internal/utils"
)

var blobsCmd = &cobra.Command{
	Use:   "blobs",
	Short: "Inspect blobs, manifests and orphans in the Ollama models directory",
	RunE: func(cmd *cobra.Command, args []string) error {
		modelsRoot, _ := cmd.Flags().GetString("models-root")
		modelsRoot = filepath.Join(os.ExpandEnv(modelsRoot))

		root := os.ExpandEnv(filepath.Join(os.Getenv("HOME"), modelsRoot))
		if !strings.HasPrefix(modelsRoot, "/") && !strings.HasPrefix(modelsRoot, "~") {
			root = filepath.Join(os.ExpandEnv("~"), modelsRoot)
		}

		manifestRoot := filepath.Join(root, "manifests", "registry.ollama.ai")
		blobsRoot := filepath.Join(root, "blobs")

		debug, _ := cmd.Flags().GetBool("debug")
		if debug {
			fmt.Printf("DEBUG: models_root   = %s\n", root)
			fmt.Printf("DEBUG: manifest_root = %s\n", manifestRoot)
			fmt.Printf("DEBUG: blobs_root    = %s\n", blobsRoot)
		}

		blobToModels := blobscan.CollectBlobMappings(manifestRoot)
		blobs := blobscan.ListAllBlobs(blobsRoot)

		if debug {
			fmt.Printf("DEBUG: total blobs found in blobs/ = %d\n", len(blobs))
		}

		type Row struct {
			Blob     string
			Models   string
			SizeB    int64
			Size     string
			IsOrphan string
		}

		var rows []Row
		var orphanFiles []string

		totalBlobs := len(blobs)
		showProgress, _ := cmd.Flags().GetBool("progress")
		if showProgress && totalBlobs > 0 {
			fmt.Printf("Processing %d blobs...\n", totalBlobs)
		}

		for idx, blobHash := range blobs {
			if showProgress && totalBlobs > 0 {
				fmt.Printf("\rBlobs: [%-30s] %d/%d", strings.Repeat("#", idx*30/totalBlobs), idx+1, totalBlobs)
			}

			blobFile := filepath.Join(blobsRoot, "sha256-"+blobHash)
			models := blobToModels[blobHash]
			sort.Strings(models)

			stat, _ := os.Stat(blobFile)
			var bsize int64
			if stat != nil {
				bsize = stat.Size()
			}

			format, _ := cmd.Flags().GetString("format")
			sizeStr := utils.FormatSize(bsize, format)

			isOrphan := len(models) == 0
			if isOrphan {
				orphanFiles = append(orphanFiles, blobFile)
			}

			rows = append(rows, Row{
				Blob:     "sha256-" + blobHash,
				Models:   strings.Join(models, "|"),
				SizeB:    bsize,
				Size:     sizeStr,
				IsOrphan: map[bool]string{true: "yes", false: "no"}[isOrphan],
			})
		}

		if showProgress && totalBlobs > 0 {
			fmt.Println()
		}

		onlyOrphans, _ := cmd.Flags().GetBool("only-orphans")
		if onlyOrphans {
			var filtered []Row
			for _, r := range rows {
				if r.IsOrphan == "yes" {
					filtered = append(filtered, r)
				}
			}
			rows = filtered
		}

		// Sorting
		sortByBlob, _ := cmd.Flags().GetBool("sort-by-blob")
		sortByModel, _ := cmd.Flags().GetBool("sort-by-model")
		sortBySize, _ := cmd.Flags().GetBool("sort-by-size")
		sortDesc, _ := cmd.Flags().GetBool("sort-desc")

		reverse := sortDesc

		sort.Slice(rows, func(i, j int) bool {
			var less bool
			if sortByBlob {
				less = rows[i].Blob < rows[j].Blob
			} else if sortByModel {
				iEmpty := rows[i].Models == ""
				jEmpty := rows[j].Models == ""
				if iEmpty != jEmpty {
					less = iEmpty
				} else {
					less = rows[i].Models < rows[j].Models
				}
			} else if sortBySize {
				less = rows[i].SizeB < rows[j].SizeB
			}
			if reverse {
				less = !less
			}
			return less
		})

		// Output
		columnsStr, _ := cmd.Flags().GetString("columns")
		columns := []string{}
		for _, c := range strings.Split(columnsStr, ",") {
			if c = strings.TrimSpace(c); c != "" {
				columns = append(columns, c)
			}
		}

		if len(columns) == 0 {
			fmt.Fprintf(os.Stderr, "Error: no columns specified\n")
			os.Exit(2)
		}

		asCSV, _ := cmd.Flags().GetBool("as-csv")
		outPath, _ := cmd.Flags().GetString("output")

		if asCSV {
			var file *os.File
			var err error
			if outPath == "-" || outPath == "" {
				file = os.Stdout
			} else {
				file, err = os.Create(outPath)
				if err != nil {
					return err
				}
				defer file.Close()
			}

			writer := csv.NewWriter(file)
			writer.Write(columns)
			for _, r := range rows {
				var record []string
				for _, col := range columns {
					switch col {
					case "blob":
						record = append(record, r.Blob)
					case "models":
						record = append(record, r.Models)
					case "size_bytes":
						record = append(record, fmt.Sprintf("%d", r.SizeB))
					case "size":
						record = append(record, r.Size)
					case "is_orphan":
						record = append(record, r.IsOrphan)
					}
				}
				writer.Write(record)
			}
			writer.Flush()
		} else {
			noColor, _ := cmd.Flags().GetBool("no-color")
			enableColor := !noColor && utils.SupportsColor()

			widths := make(map[string]int)
			for _, col := range columns {
				widths[col] = len(col)
			}
			for _, r := range rows {
				for _, col := range columns {
					var val string
					switch col {
					case "blob":
						val = r.Blob
					case "models":
						val = r.Models
					case "size":
						val = r.Size
					case "is_orphan":
						val = r.IsOrphan
					}
					if len(val) > widths[col] {
						widths[col] = len(val)
					}
				}
			}

			header := ""
			for i, col := range columns {
				if i > 0 {
					header += "  "
				}
				header += padRight(col, widths[col])
			}
			fmt.Println(utils.Colorize(header, "1;37", enableColor))

			for _, r := range rows {
				line := ""
				for i, col := range columns {
					if i > 0 {
						line += "  "
					}
					var val string
					switch col {
					case "blob":
						val = r.Blob
					case "models":
						val = r.Models
					case "size":
						val = r.Size
					case "is_orphan":
						val = r.IsOrphan
					}
					val = padRight(val, widths[col])
					if r.IsOrphan == "yes" {
						val = utils.Colorize(val, "31", enableColor)
					}
					line += val
				}
				fmt.Println(line)
			}
		}

		// Delete orphans
		deleteOrphans, _ := cmd.Flags().GetBool("delete-orphans")
		if deleteOrphans {
			if len(orphanFiles) == 0 {
				fmt.Println("\nNo orphan blobs found.")
				return nil
			}

			fmt.Println("\nOrphan blobs to delete:")
			for _, f := range orphanFiles {
				fmt.Printf("  blobs/%s\n", filepath.Base(f))
			}

			force, _ := cmd.Flags().GetBool("force")
			if !force {
				fmt.Print("Delete? (yes/no) ")
				var confirm string
				fmt.Scanln(&confirm)
				if strings.ToLower(confirm) != "yes" {
					fmt.Println("Aborted.")
					return nil
				}
			}

			for _, f := range orphanFiles {
				if err := os.Remove(f); err == nil {
					fmt.Printf("Deleted: blobs/%s\n", filepath.Base(f))
				} else {
					fmt.Printf("Failed: %s %v\n", f, err)
				}
			}
			fmt.Println("Done.")
		}

		return nil
	},
}

func init() {
	blobsCmd.Flags().String("models-root", "~/.ollama/models", "")
	blobsCmd.Flags().Bool("as-csv", false, "Output as CSV")
	blobsCmd.Flags().StringP("output", "o", "-", "CSV output path (default: stdout)")
	blobsCmd.Flags().Bool("delete-orphans", false, "Delete orphan blobs")
	blobsCmd.Flags().Bool("force", false, "Do not ask before deleting")
	blobsCmd.Flags().Bool("only-orphans", false, "Show only orphans")
	blobsCmd.Flags().Bool("sort-by-blob", false, "Sort by blob name")
	blobsCmd.Flags().Bool("sort-by-model", false, "Sort by first model name")
	blobsCmd.Flags().Bool("sort-by-size", false, "Sort by size in bytes")
	blobsCmd.Flags().Bool("sort-asc", false, "Sort ascending (default)")
	blobsCmd.Flags().Bool("sort-desc", false, "Sort descending")
	blobsCmd.Flags().String("format", "mb", "Format for human-readable size (mb, gb)")
	blobsCmd.Flags().Bool("no-color", false, "Disable colored output")
	blobsCmd.Flags().String("columns", "blob,models,size,is_orphan", "Comma-separated list of columns")
	blobsCmd.Flags().Bool("progress", false, "Show a simple progress bar while processing blobs")
	blobsCmd.Flags().Bool("debug", false, "Print debug information (paths, blob count, etc.)")
	blobsCmd.Flags().String("debug-blob", "", "Hash (without 'sha256-') to debug presence")
}
